File: html5/RadXml/RadXmlEvents.js
/*
	MIT-LICENSE
	Copyright (c) 2013 Higher Edge Software, LLC

	Permission is hereby granted, free of charge, to any person obtaining a copy of this software 
	and associated documentation files (the "Software"), to deal in the Software without restriction, 
	including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, 
	and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, 
	subject to the following conditions:

	The above copyright notice and this permission notice shall be included in all copies or substantial 
	portions of the Software.

	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT 
	LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. 
	IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
	LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION 
	WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

function Event (name, javascript)
{
	if (name == null)
		name = "";

	if (javascript == null)
		javascript = "";

	this._name = name;
	this._javascript = javascript;

	this._setName = function (newName)
	{
		this._name = newName;
	}

	this._getName = function ()
	{
		return (name);
	}

	this._setJavascript = function (newJavascript)
	{
		this._javascript = newJavascript;
	}

	this._getJavascript = function ()
	{
		return (this._javascript);
	}
}

function Events ()
{
	this._hshEvents = new Object ();

	this._addEvent = function (newEvent)
	{
		this._hshEvents[newEvent._getName ()] = newEvent;
	}

	this._getEvent = function (findName)
	{
		return (this._hshEvents[findName]);
	}

	this._setJavascriptToEvent = function (findName, code)
	{
		this._hshEvents[findName]._javascript = code;
	}

	this._getJavascriptFromEvent = function (findName)
	{
		return (this._hshEvents[findName]._javascript);
	}

	this._executeJavascriptFromEvent = function (findName)
	{
		var strJS = this._hshEvents[findName]._javascript;
		RadXML.compileAndRunJSString (strJS);
	}
}

File: source/RadXml/RadXmlEvents.cpp
#include "RadXmlEvents.h"

#include "RadXml.h"

#include "common/com_radxml_strconv.h"

namespace RadXml
{
	// Event
	Event::Event (std::string newName, std::string newJavascript)
	{
		setName (newName);
		setJavascript (newJavascript);
	}

	void Event::setName (std::string newName)
	{
		name = newName;
		hash = RadXmlCommon::s13StringToHash (newName);
	}

	// Events
	Events::Events ()
	{
	}

	Events::Events (const Events &copy)
	{
		HASHMAP<unsigned long, Event *>::const_iterator itBegin = copy.hshEvents.begin ();
		HASHMAP<unsigned long, Event *>::const_iterator itEnd = copy.hshEvents.end ();

		while (itBegin != itEnd)
		{
			Event *evtEvent = (*itBegin).second;
			Event *evtClone = new Event (*evtEvent);

			hshEvents.insert (std::pair<unsigned long, Event *> (evtClone->getHash (), evtClone));

			itBegin++;
		}
	}

	Events::~Events ()
	{
		HASHMAP<unsigned long, Event *>::iterator itBegin = hshEvents.begin ();
		HASHMAP<unsigned long, Event *>::iterator itEnd = hshEvents.end ();

		while (itBegin != itEnd)
		{
			Event *evtEvent = (*itBegin).second;

			if (evtEvent != 0)
			{
				delete evtEvent;
				evtEvent = 0;
			}

			itBegin++;
		}
	}

	void Events::addEvent (Event *newEvent)
	{
		hshEvents.insert (std::pair<unsigned long, Event *> (newEvent->getHash (), newEvent));
	}

	Event *Events::getEvent (std::string findName)
	{
		unsigned long ulHash = RadXmlCommon::s13StringToHash (findName);
		HASHMAP<unsigned long, Event *>::iterator itBegin = hshEvents.find (ulHash);
		HASHMAP<unsigned long, Event *>::iterator itEnd = hshEvents.end ();

		if (itBegin != itEnd)
			return ((*itBegin).second);

		return (0);
	}

	Event *Events::getEvent (unsigned long findHash)
	{
		HASHMAP<unsigned long, Event *>::iterator itBegin = hshEvents.find (findHash);
		HASHMAP<unsigned long, Event *>::iterator itEnd = hshEvents.end ();

		if (itBegin != itEnd)
			return ((*itBegin).second);

		return (0);
	}

	void Events::setJavascriptToEvent (std::string findName, std::string javascript)
	{
		Event *evtEvent = getEvent (findName);
		evtEvent->setJavascript (javascript);
	}

	void Events::setJavascriptToEvent (unsigned long findHash, std::string javascript)
	{
		Event *evtEvent = getEvent (findHash);
		evtEvent->setJavascript (javascript);
	}

	std::string Events::getJavascriptFromEvent (std::string findName)
	{
		Event *evtEvent = getEvent (findName);

		if (evtEvent == 0)
			return ("");

		return (evtEvent->getJavascript ());
	}

	std::string Events::getJavascriptFromEvent (unsigned long findHash)
	{
		Event *evtEvent = getEvent (findHash);

		if (evtEvent == 0)
			return (0);

		return (evtEvent->getJavascript ());
	}

	bool Events::executeJavascriptFromEvent (std::string findName)
	{
		RadXML *ogrGame = RadXML::getSingletonPtr ();
		std::string javascript = getJavascriptFromEvent (findName);

		if (javascript == "")
			return (true);

		return (ogrGame->compileAndRunJSString (javascript));
	}

	bool Events::executeJavascriptFromEvent (unsigned long findHash)
	{
		std::string javascript = getJavascriptFromEvent (findHash);

		if (javascript == "")
			return (true);

		RadXML *ogrGame = RadXML::getSingletonPtr ();

		return (ogrGame->compileAndRunJSString (javascript));
	}
}

